qiskit/transpiler/target.py

```diff
@@ -0,0 +1,367 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+A target object represents the minimum set of information the transpiler needs
+from a backend
+"""
+
+import logging
+
+import retworkx as rx
+
+from qiskit.transpiler.coupling import CouplingMap
+from qiskit.transpiler.exceptions import CouplingError
+from qiskit.transpiler.instruction_durations import InstructionDurations
+
+logger = logging.getLogger(__name__)
+
+
+class InstructionProperties:
+    """A representation of the properties of a gate implementation."""
+
+    __slots__ = ("length", "error", "properties")
+
+    def __init__(self, length=None, error=None, properties=None):
+        self.length = length
+        self.error = error
+        self.properties = properties
+
+
+class Target:
```

Comment:
It doesn't look like this supports different coupling maps based on the type of 2Q gate.  Perhaps I am not remembering properly, but I recall that a system might have a collection of different 2Q gates available to it, but that they might correspond to different connectivity; not all gates need to calibrated for all edges of the physical coupling map.  This is why some time ago it was suggested that the `coupling_map` be the actual device connectivity that could in principle be targeted, e.g. via pulse, where as the `gate_map` would be the connectivity supported by that particular 2Q gate.  But again, it has been a while since I last thought about this.

Reply:
I don't have a method for it but the data is all there in the class. Right now the `coupling_map()` method returns the full combined coupling map with all 2q instructions in the target. But, I created this data structure with heterogeneous 2q gates in mind so we can represent some edges only exist with certain gates (the goal being to update the transpiler in a follow up to use this with BackendV2. But I agree it'll be useful to have this so I'll add a kwarg to the coupling map method to optionally specify a 2q gate and use that to return just a coupling map object for that gate.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5885/commits/fd571466a96118034cfa13df2ca8d82980e2d6e1